synapse/storage/util/id_generators.py

```diff
@@ -101,7 +100,10 @@ def __init__(self, db_conn, table, column, extra_tables=[], step=1):
             self._current = (max if step > 0 else min)(
                 self._current, _load_current_id(db_conn, table, column, step)
             )
-        self._unfinished_ids = deque()  # type: Deque[int]
+
+        # We use this as an ordered set, as we want to efficiently append items,
+        # remove items and get the first item.
+        self._unfinished_ids = OrderedDict()  # type: OrderedDict[int, int]
```

Comment:
`OrderedDict` is ordered by insertion order rather than key, iirc. Is that what you intend here? (possibly it amounts to the same thing?)

Reply:
Yeah, they amount to the same thing (and is the same situation as with deque). Will expand the comment